### PR TITLE
feature: client vscode extension for xml LSP server 

### PR DESCRIPTION
**Pull request by rayakoren**

### CODE REVIEW

File: packages/xml-for-vscode/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# LSP Example
```
Comment (bd82) [inline packages/xml-for-vscode/README.md]: README from sample does not belong here.

File: packages/xml-for-vscode/package.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "vscode-xml-tools-client",
```
Comment (bd82) [inline packages/xml-for-vscode/package.json]: xml-for-vscode name?

File: packages/xml-for-vscode/package.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  "description": "xml-tools for VS Code",
```
Comment (bd82) [inline packages/xml-for-vscode/package.json]: VSCode is a single word

File: packages/xml-for-vscode/src/extension.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import { workspace, ExtensionContext } from "vscode";
```
Comment (bd82) [inline packages/xml-for-vscode/src/extension.ts]: convert to JavaScript

File: packages/xml-for-vscode/package.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    "vscode:prepublish": "npm run update-vscode && cd .. && npm run compile",
+    "postinstall": "vscode-install",
+    "compile": "tsc -b",
+    "watch": "tsc -b -w",
+    "build": "vsce package",
+    "test": "tsc -b && node ./out/test/runTest.js"
+  },
+  "dependencies": {
+    "@xml-tools/language-server": "1.0.2",
+    "vscode-languageclient": "^5.2.1"
```
Comment (bd82) [inline packages/xml-for-vscode/package.json]: upgrade to version 6?

Comment (bd82) [reply packages/xml-for-vscode/package.json]: used fixed version

File: packages/xml-for-vscode/src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  LanguageClientOptions,
+  ServerOptions,
+  TransportKind
+} from "vscode-languageclient";
+import { SERVER_PATH } from "@xml-tools/language-server";
+
+let client: LanguageClient;
+
+export function activate(context: ExtensionContext) {
+  // The debug options for the server
```
Comment (bd82) [inline packages/xml-for-vscode/src/extension.ts]: remove redundant example comments

File: packages/xml-for-vscode/src/extension.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    synchronize: {
+      // Notify the server about file changes to '.xml files contained in the workspace
+      fileEvents: workspace.createFileSystemWatcher("**/*.xml")
+    }
+  };
+
+  // Create the language client and start the client.
+  client = new LanguageClient(
+    "XMLtools",
+    "XML Tools Client",
```
Comment (bd82) [inline packages/xml-for-vscode/src/extension.ts]: these names have changed

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+const docUri = getDocUri(docPath);
+
+describe("Extension Test", () => {
+  before(async function() {
+    await vscode.workspace.openTextDocument(docUri);
+    await activate();
+    //wait for extension to load
+    await sleep(2000);
+  });
+  it("Test diagnostic text - lexing errors", async function() {
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: another line terminator before `it`

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
+import * as path from "path";
+import { getDocUri, activate, sleep } from "./helper";
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: relative imports **after** packages imports

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+describe("Extension Test", () => {
+  before(async function() {
+    await vscode.workspace.openTextDocument(docUri);
+    await activate();
+    //wait for extension to load
+    await sleep(2000);
+  });
+  it("Test diagnostic text - lexing errors", async function() {
+    let content = `<note!>
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: const

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+  <to>Tove</to>
+  <from>Jani</from>
+  <heading>Reminder</heading>
+  <body>Don't forget me this weekend!</body>
+</note>`;
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: format the xml snippet manually so it would be better aligned with rest of JS code

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    await fs.writeFile(docPath, content);
+    await sleep(2000);
+    await vscode.window.showTextDocument(docUri);
+  }
+
+  async function testDiagnostics(expectedDiagnostics: vscode.Diagnostic[]) {
+    let actualDiagnostics: vscode.Diagnostic[] = vscode.languages.getDiagnostics(
+      docUri
+    );
+    console.log(JSON.stringify(actualDiagnostics));
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: remove console.log

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+  async function setContent(content: string) {
+    await vscode.workspace.openTextDocument(docUri);
+    await fs.writeFile(docPath, content);
+    await sleep(2000);
+    await vscode.window.showTextDocument(docUri);
+  }
+
+  async function testDiagnostics(expectedDiagnostics: vscode.Diagnostic[]) {
+    let actualDiagnostics: vscode.Diagnostic[] = vscode.languages.getDiagnostics(
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: const.

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+      docUri
+    );
+    console.log(JSON.stringify(actualDiagnostics));
+    console.log(JSON.stringify(expectedDiagnostics));
+    expect(actualDiagnostics.length).to.equal(expectedDiagnostics.length);
+
+    expectedDiagnostics.forEach((expectedDiagnostic, i) => {
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: - try to use deep.equals or include.members instead of iterating over expected diagnostics.
- try to avoid assumptions on order (include.members > deep.equals)

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+  function toRange(sLine: number, sChar: number, eLine: number, eChar: number) {
+    const start = new vscode.Position(sLine, sChar);
+    const end = new vscode.Position(eLine, eChar);
+    return new vscode.Range(start, end);
+  }
+
+  async function setContent(content: string) {
+    await vscode.workspace.openTextDocument(docUri);
+    await fs.writeFile(docPath, content);
+    await sleep(2000);
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: tests fail without this sleep?

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        severity: vscode.DiagnosticSeverity.Error,
+        message: "Expecting token of type --> CLOSE <-- but found --> '' <--",
+        range: toRange(5, null, 0, 0),
+        source: SYNTAX_ERROR_MSG
+      }
+    ]);
+  });
+
+  it("Test diagnostic text - Multiple errors", async function() {
+    let content = `<note>
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: const

File: packages/xml-for-vscode/src/test/suite/extension.test.ts [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+  "..",
+  "src",
+  "test",
+  "testFixure",
+  "test.xml"
+);
+const docUri = getDocUri(docPath);
+
+describe("Extension Test", () => {
+  before(async function() {
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/extension.test.ts]: More **encapsulation** of tests desired.
Possibilities:
- add afterEach to clean the content? (I think this is better)
- open a different for each test?

File: packages/xml-for-vscode/src/test/suite/helper.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+export let doc: vscode.TextDocument;
+export let editor: vscode.TextEditor;
+export let documentEol: string;
+export let platformEol: string;
+
+/**
+ * Activates the extension 'vscode.vscode-xml-tools-client'
+ */
+export async function activate() {
+  let ext = vscode.extensions.getExtension("SAPSE.vscode-xml-tools-client");
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/helper.ts]: let there be const

File: packages/xml-for-vscode/src/test/suite/helper.ts [hunk 19]
```diff
@@ -0,0 +1,22 @@
+import * as vscode from "vscode";
+
+export let doc: vscode.TextDocument;
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/helper.ts]: consts...

File: packages/xml-for-vscode/src/test/suite/helper.ts [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+export let editor: vscode.TextEditor;
+export let documentEol: string;
+export let platformEol: string;
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/suite/helper.ts]: why are we exporting empty variables which are never assigned?

File: packages/xml-for-vscode/src/test/testFixure/test.xml [hunk 21]
```diff
@@ -0,0 +1,6 @@
+<note>
```
Comment (bd82) [inline packages/xml-for-vscode/src/test/testFixure/test.xml]: if we do "setContent" all the time do we need initial content?

File: packages/xml-for-vscode/package.json [hunk 22]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "vscode-xml-tools-client",
+  "description": "xml-tools for VS Code",
+  "version": "0.0.1",
```
Comment (bd82) [inline packages/xml-for-vscode/package.json]: needs "private:" : true to avoid publishing to npm

File: packages/xml-for-vscode/lib/extension.js [hunk 23]
```diff
@@ -0,0 +1,46 @@
+const { workspace } = require("vscode");
+const { LanguageClient, TransportKind } = require("vscode-languageclient");
+const { SERVER_PATH } = require("@xml-tools/language-server");
+
+let client;
+
+function activate() {
+  let debugOptions = { execArgv: ["--nolazy", "--inspect=6009"] };
```
Comment (bd82) [inline packages/xml-for-vscode/lib/extension.js]: `let` -> `const`

File: packages/xml-for-vscode/nyc.config.js [hunk 24]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
```
Comment (bd82) [inline packages/xml-for-vscode/nyc.config.js]: - 100%* coverage
- https://reflectoring.io/100-percent-test-coverage/
- don't use an nyc.config.js

File: packages/xml-for-vscode/package.json [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+    "url": "https://github.com/SAP/xml-tools"
+  },
+  "engines": {
+    "vscode": "1.42.1"
+  },
+  "activationEvents": [
+    "onLanguage:xml"
+  ],
+  "main": "lib/extension",
+  "scripts": {
```
Comment (bd82) [inline packages/xml-for-vscode/package.json]: align scripts with other packages structure

File: packages/xml-for-vscode/lib/extension.js [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    "XML For VSCode",
+    serverOptions,
+    clientOptions
+  );
+
+  client.start();
+}
+
+function deactivate() {
+  if (!client) {
```
Comment (bd82) [inline packages/xml-for-vscode/lib/extension.js]: can ignore this branch (edge case)

File: packages/xml-for-vscode/package.json [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+    "test": "node test/runTest.js",
+    "coverage": "nyc node test/runTest.js",
+    "build": "vsce package"
+  },
+  "dependencies": {
+    "@xml-tools/language-server": "1.0.2",
+    "vscode-languageclient": "6.1.1"
+  },
+  "devDependencies": {
+    "@types/node": "10.17.15",
```
Comment (bd82) [inline packages/xml-for-vscode/package.json]: extract node types deps to top level package.

File: packages/xml-for-vscode/test/suite/extension.test.js [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+describe("XML for VSCode extension", () => {
+  before(async function() {
+    await vscode.workspace.openTextDocument(docUri);
+    await vscode.window.showTextDocument(docUri);
+    // We need to explicitly wait for extension to load
+    await sleep(1000);
+  });
+
+  after(async function() {
+    // Clear the xml document
```
Comment (bd82) [inline packages/xml-for-vscode/test/suite/extension.test.js]: comments may be redundant here

File: packages/xml-for-vscode/test/suite/extension.test.js [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+    await sleep(1000);
+  }
+
+  async function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+
+  async function testDiagnostics(docUri, expectedDiagnostics) {
+    const actualDiagnostics = vscode.languages.getDiagnostics(docUri);
+    expect(expectedDiagnostics).to.deep.equal(actualDiagnostics);
```
Comment (bd82) [inline packages/xml-for-vscode/test/suite/extension.test.js]: use: https://www.chaijs.com/plugins/deep-equal-in-any-order/

It will avoid tests breaking due to possible changes in language-server implementation
